msal/application.py

```diff
@@ -405,7 +405,7 @@ def _acquire_token_by_username_password_federated(
             raise RuntimeError(
                 "RSTR returned unknown token type: %s", wstrust_result.get("type"))
         return self.client.obtain_token_by_assertion(
-            b64encode(wstrust_result["token"]),
+            urlsafe_b64encode(wstrust_result["token"]).strip(b'='),
```

Comment:
This turns out becoming an interesting journey.

Yes `rstrip(...)` would be enough, and now I use it in the underlying implementation. Thanks for the suggestion!

And then it becomes the last straw that I realize we should not have forced the caller to do such tricky ad-hoc encoding in the first place (even though we documented it). We should have wrapped it inside a helper instead, and do the right behavior by default. Which is what we do now in this updated PR. And, as a byproduct, the grant_type parameter is no longer optional, because we removed that hacky `if b"." in assertion` thing. None of the above boring details matters in the MSAL public API `acquire_token_by_username_password()`, because this PR is just implementation details.

I'll merge this PR soon. Happy refactoring!